R/complete_magpie.R

```diff
@@ -1,40 +1,52 @@
 #' complete_magpie
-#' 
+#'
 #' MAgPIE objects can be incomplete to reduce memory. This function blows up a
 #' magpie object to its real dimensions, so you can apply unwrap.
-#' 
-#' 
+#'
+#'
 #' @param x MAgPIE object which should be completed.
 #' @param fill Value that shall be written into the missing entries
+#' @param dim dimensions in which the completion should take place (1, 2 and/or 3). For
+#' full completion use \code{1:3}
 #' @return The completed MAgPIE object
-#' @author Benjamin Bodirsky
+#' @author Jan Philipp Dietrich, Benjamin Bodirsky
 #' @seealso \code{\link{add_dimension}},\code{\link{clean_magpie}}
 #' @examples
-#' 
-#'  pop <- maxample("pop")
-#'  a <- complete_magpie(pop)
-#'  b <- add_dimension(a)
-#'  c <- add_dimension(a,nm="dummy2")
-#'  incomplete<-mbind(b[,,1],c)
-#'  d<-complete_magpie(incomplete)
-#' 
-#' @export complete_magpie
-complete_magpie<-function(x,fill=NA) {
-  full<-fulldim(x)[[2]]
-  permute<-full[[3]]
-  repeatit<-length(permute)
-  if (length(full)>3) {
-    for (i in 4:length(full)) {
-      permute<-paste(rep(permute,each=length(full[[i]])),full[[i]],sep=".")
-      repeatit<-length(permute)
-    }
+#'
+#' pop <- maxample("pop")
+#' complete_magpie(pop)
+#'
+#' ani <- maxample("animal")
+#' complete_magpie(ani)
+#' @export
+
+complete_magpie <- function(x, fill = NA, dim = 3) { #nolint
+  .expand <- function(x) {
+    grid <- expand.grid(x, stringsAsFactors = FALSE)
+    return(sort(apply(grid, 1, paste, collapse = ".")))
   }
-  missing<-permute[!(permute%in%dimnames(x)[[3]])]
-  if(length(missing)>0){
-    add<-new.magpie(cells_and_regions = full[[1]],years = full[[2]],names = missing,fill=fill)
-    out<-mbind(x,add)
-  } else {out<-x}
-  out<-out[,,order(getNames(out))]
-  getMetadata(out) <- getMetadata(x)
+  .completeSize <- function(x, dimIn, dim) {
+    .prod <- function(x) return(prod(vapply(x, length, 1)))
```

Comment:
```suggestion
    .prod <- function(x) return(prod(vapply(x, length, numeric(1))))
```

Reply:
This is just semantic sugar, but I like writing out the return type for `vapply`.